### PR TITLE
Issue 3020: Updating reader-group.md

### LINKED ISSUE #3020: Request to update content: reader-group-design.md
**Problem description**
1. To be updated to reflect recent implementation details (05/07/2017)

**Problem location**
reader-group-design. md

### PR DESCRIPTION
**Pull request by Aparnarr**
**Change log description**  
1. Format issues fixed
2. Minor content edition performed.
3. Hyperlinks checked
4. **Request the developers (Sandeep and Tom) to check for any inclusion.**

**Purpose of the change**  
Fixes #3020

### CODE REVIEW

File: documentation/src/docs/reader-group-design.md [hunk 1]
```diff
@@ -11,83 +11,85 @@ You may obtain a copy of the License at
 
 _To be updated to reflect recent implementation details (05/07/2017)_
 
-## Motivation 
-A set of Readers can be grouped together in order that the set of Events in a Stream can be read in parallel.  The grouping of Readers is called a Readers Group.  Pravega guarantees that each Event in the Stream is read by exactly one Reader in the Readers Group. 
+## Motivation
+A set of Readers can be grouped together in order that the set of Events in a Stream can be read in parallel. This grouping of Readers is called a Reader Group. Pravega guarantees that each Event in the Stream is read by exactly one Reader in the Reader Group.
```
Comment (shrids) [inline documentation/src/docs/reader-group-design.md]: My understanding of the pattern followed for documentation is as follows.
a. If we are referring to a concept then the naming convention is as `ReaderGroup` i.e. within \` \`
b. if we are talking about a stream in general term then it can start with lower case and without \` \`

Hence we could consistently name Reader Group as `ReaderGroup` , similarly for readers based on context.

File: documentation/src/docs/reader-group-design.md [hunk 2]
```diff
@@ -11,83 +11,85 @@ You may obtain a copy of the License at
...
-- How to manage the above mapping when Segments split and merge
-- How to manage the above mapping when Readers are added to the ReaderGroup 
-- How to manage the above mapping when Readers leave the ReaderGroup either by an explicit operation or the Reader becoming unavailable due to network outage or the Reader process failing
+ -  How to maintain the mapping of which Reader within a Reader Group is assigned which Segment?
+ -  How to manage the above mapping when Segments split and merge?
+ -  How to manage the above mapping when Readers are added to the Reader Group?
+ -  How to manage the above mapping when Readers leave the Reader Group either by an explicit operation or the Reader becoming unavailable due to network outage or the Reader process failing?
 
-To solve these problems we can use [StateSynchronizer](state-synchronizer-design.md) to enable readers to coordinate among themselves.
+To address these challenges, we can use [State Synchronizer](state-synchronizer-design.md) to enable coordination among Readers.
```
Comment (shrids) [inline documentation/src/docs/reader-group-design.md]: Here `StateSynchronizer` is a single word.

File: documentation/src/docs/reader-group-design.md (hunk 1 above)
Comment (Aparnarr) [reply documentation/src/docs/reader-group-design.md]: Hi @shrids ,
will go with the terminology document. In this situation we refer as concept, so no changes required!

File: documentation/src/docs/reader-group-design.md [hunk 3]
```diff
@@ -11,83 +11,85 @@ You may obtain a copy of the License at
 
 _To be updated to reflect recent implementation details (05/07/2017)_
```
Comment (Aparnarr) [inline documentation/src/docs/reader-group-design.md]: @shrids  Any further updation required?

File: documentation/src/docs/reader-group-design.md (hunk 2 above)
Comment (Aparnarr) [reply documentation/src/docs/reader-group-design.md]: Hi @RaulGracia,
As we discussed, we need to decide on the naming conventions... (Reader Group, Reader/Readers).
Hope in this docs the terms are edited correctly?

Comment (Aparnarr) [reply documentation/src/docs/reader-group-design.md]: @Aparnarr

Comment (Aparnarr) [reply documentation/src/docs/reader-group-design.md]: @fpj Please include @RaulGracia  as another reviewer

File: documentation/src/docs/reader-group-design.md (hunk 3 above)
Comment (shrids) [reply documentation/src/docs/reader-group-design.md]: yes, please find the summary below
- Reader creation:
   The document talks about a reader creation using the snippet `a_stream.createReader(...)` this has to be replaced with the snippet
```
ClientFactory clientFactory = ClientFactory.withScope(scope, controllerURI);
EventStreamReader<Integer> reader = clientFactory.createReader(readerId, READER_GROUP_NAME, serializer, readerConfig);
```
(for reference use: https://git.io/fxStL)

- Reader going offline (section)
Here the documentation talks about how the last known position of the reader is passed to the `readerOffline(readerId, position) api`.
If the position is sent as "null" then the last checkpointed position of the reader is used by the newer readers when they take ownership of the segment(s) that were read by the older/offline reader.

Comment (Aparnarr) [reply documentation/src/docs/reader-group-design.md]: can i mention position object or parameter

File: documentation/src/docs/reader-group-design.md [hunk 4]
```diff
@@ -79,12 +78,16 @@ There are no races between multiple Readers coming online concurrently because o
 
 There is no ambiguity as to who the owner is, because it is stored in the shared state. There is no risk of a segment being ignored because every Reader can see the available segments by looking at the shared state and claim them.
 
-## Reader going offline.
-1. When a Reader dies, the `ReaderOffline()` method will be invoked either by the Reader itself in a graceful shutdown (internally to the close method) or via a "liveness detector". In either case the Reader's last position is written to the state.
-1. Other Readers will see this when they update their local state.
+## Reader going offline
+1. When a Reader dies, the `ReaderOffline()` method will be invoked either by the Reader itself in a graceful shutdown (internally to the close method) or via a "liveness detector". In either case the Reader's last position is written to the state `readerOffline(readerId, position) api`.
```
Comment (shrids) [inline documentation/src/docs/reader-group-design.md]: readerOffline(readerId, position) method starts with lower case, right? 
We could replace `ReaderOffline()` with the actual api.
The second `readerOffline` reference seemed redundant.

File: documentation/src/docs/reader-group-design.md [hunk 5]
```diff
@@ -9,85 +9,90 @@ You may obtain a copy of the License at
...
+1. When a Reader comes to the end of its segment it records this information in the shared state.
+1. When all of the segments that are getting merged together are completed, a Reader may claim ownership of the following segment.
+
+There is no ambiguity as to who the owner is, because it is stored in the shared state. There is no risk of a segment being ignored because every Reader can see the available segments by looking at the shared state and claim them.
+
+## Reader going offline
+1. When a Reader dies, the `ReaderOffline()` method will be invoked either by the Reader itself in a graceful shutdown (internally to the close method) or via a "liveness detector". In either case the Reader's last position is written to the state `readerOffline(readerId, position) api`.
+
+1. Other Readers will see this `position` object when they update their local state.
+1. If the `position` is sent as "null" then the last checkpointed position of the Reader is used by the newer Readers when they take ownership of the segment(s) that were read by the older/offline reader.
```
[the remark this answers is not in the document]
Comment (shrids) [reply documentation/src/docs/reader-group-design.md]: The standard naming pattern should be followed i.e. `Position` or position. 
Perhaps we could simplify the wording as 
> (If a `null` `Position` is sent then the last checkpointed position will be written to the state.)

[the remark this answers is not in the document]
Comment (Aparnarr) [reply documentation/src/docs/reader-group-design.md]: Accepted Suggestion

File: documentation/src/docs/reader-group-design.md (hunk 4 above)
Comment (Aparnarr) [reply documentation/src/docs/reader-group-design.md]: Accepted suggestion

File: documentation/src/docs/reader-group-design.md [hunk 6]
```diff
@@ -52,8 +50,9 @@ The external APIs to manage Reader Groups could be added to the `StreamManager`
 When a Reader Group is created, it creates a [State Synchronizer](state-synchronizer-design.md) shared by the Readers. To join a Reader Group, Readers would just specify it in their configuration:
 
 ```
-    ReaderConfig cc = new ReaderConfig(props);
-    Reader<T> Reader = a_stream.createReader("my_Reader_id", "my_Reader_group", cc);
+ClientFactory clientFactory = ClientFactory.withScope(scope, controllerURI);
+EventStreamReader<Integer> reader = clientFactory.createReader(readerId, READER_GROUP_NAME, serializer, readerConfig);
```
[the remark this answers is not in the document]
Comment (Aparnarr) [reply documentation/src/docs/reader-group-design.md]: Accepted suggestion

File: documentation/src/docs/reader-group-design.md [hunk 7]
```diff
@@ -9,85 +9,90 @@ You may obtain a copy of the License at
...
-_To be updated to reflect recent implementation details (05/07/2017)_
+## Motivation
+A set of Readers can be grouped together in order that the set of Events in a Stream can be read in parallel. This grouping of Readers is called a Reader Group. Pravega guarantees that each Event in the Stream is read by exactly one Reader in the Reader Group.
 
-## Motivation 
-A set of Readers can be grouped together in order that the set of Events in a Stream can be read in parallel.  The grouping of Readers is called a Readers Group.  Pravega guarantees that each Event in the Stream is read by exactly one Reader in the Readers Group. 
-
-Each Reader in a ReaderGroup is assigned zero or more Segments.
-The Reader assigned to a Segment is the only Reader within the ReaderGroup that reads Events from that Segment.  This is the fundamental mechanism by which Pravega makes ordering guarantees of Event delivery to a Reader – a Reader will receive Events in the order they were published into a Segment.
+Each Reader in a Reader Group is assigned zero or more Segments.
```
Comment (RaulGracia) [inline documentation/src/docs/reader-group-design.md]: We have used before `Stream Segments` instead of `Segments`, so we may just stick to that name.

File: documentation/src/docs/reader-group-design.md [hunk 8]
```diff
@@ -9,85 +9,90 @@ You may obtain a copy of the License at
...
+The Reader assigned to a Segment is the only Reader within the Reader Group that reads Events from that Segment. This is the fundamental mechanism by which Pravega makes ordering guarantees of Event delivery to a Reader – a Reader will receive Events in the order they were published into a Segment.
```
Comment (RaulGracia) [inline documentation/src/docs/reader-group-design.md]: I would rather prefer `written` instead of `published` as this is a storage system (different from MOM systems).

File: documentation/src/docs/reader-group-design.md [hunk 9]
```diff
@@ -9,85 +9,90 @@ You may obtain a copy of the License at
...
-- How to manage the above mapping when Segments split and merge
-- How to manage the above mapping when Readers are added to the ReaderGroup 
-- How to manage the above mapping when Readers leave the ReaderGroup either by an explicit operation or the Reader becoming unavailable due to network outage or the Reader process failing
+ -  How to maintain the mapping of which Reader within a Reader Group is assigned which Segment?
+ -  How to manage the above mapping when Segments split and merge?
+ -  How to manage the above mapping when Readers are added to the Reader Group?
+ -  How to manage the above mapping when Readers leave the Reader Group either by an explicit operation or the Reader becoming unavailable due to network outage or the Reader process failing?
 
-To solve these problems we can use [StateSynchronizer](state-synchronizer-design.md) to enable readers to coordinate among themselves.
+To address these challenges, we can use [State Synchronizer](state-synchronizer-design.md) to enable coordination among Readers.
```
Comment (RaulGracia) [inline documentation/src/docs/reader-group-design.md]: Please, change `we can use` -> `we use`. We actually do that.

File: documentation/src/docs/reader-group-design.md [hunk 10]
```diff
@@ -9,85 +9,90 @@ You may obtain a copy of the License at
...
+
 ```
-    ReaderConfig cc = new ReaderConfig(props);
-    Reader<T> reader = a_stream.createReader("my_reader_id", "my_reader_group", cc);
+ClientFactory clientFactory = ClientFactory.withScope(scope, controllerURI);
+EventStreamReader< T > reader = clientFactory.createReader(readerId, READER_GROUP_NAME, serializer, readerConfig);
+
 ```
-When readers join the group they use the state to determine which segments to read from. When they shut down they update the state so that other readers can take over their segments.
+The Readers, while joining the group access the information stored on the state to determine which segments to read from. Once when they shut down, they update the state so that other Readers can take over their segments.
```
Comment (RaulGracia) [inline documentation/src/docs/reader-group-design.md]: Comma missing: `The Readers, while joining the group(,) access the information stored on.. `

Comment (RaulGracia) [inline documentation/src/docs/reader-group-design.md]: Please, rephrase second sentence: `When a Reader leaves the group, it updates the state so that other Readers can take over its Stream Segments.`

File: documentation/src/docs/reader-group-design.md [hunk 11]
```diff
@@ -9,85 +9,90 @@ You may obtain a copy of the License at
...
 
 # Failure detector
-We still need some sort of heartbeating mechanism to tell if Readers are alive. The problem is greatly simplified because it need not produce a view of the cluster or manage any state. The component would just need to detect failures and invoke the <code>void readerOffline(String readerId, Position lastPosition);</code> api on the ReaderGroup
+
+We still need some effective mechanism to identify, whether Readers are alive. The problem is greatly simplified because it need not produce a view of the cluster or manage any state. The component would just need to detect failures and invoke the `void ReaderOffline(String ReaderId, Position lastPosition);` API on the Reader Group.
```
Comment (RaulGracia) [inline documentation/src/docs/reader-group-design.md]: Please, rephrase: `We still need some effective mechanism to identify whether Readers are alive or not.`

File: documentation/src/docs/reader-group-design.md [hunk 12]
```diff
@@ -9,85 +9,90 @@ You may obtain a copy of the License at
...
 
 # Examples
-## New reader
-1. When a reader joins a group its online status is added to the shared state
-1. Other readers receive updates to the shared state.
-1. When a reader with more than average number of segments sees the new reader, it may give up a segment by writing its position for that segment to the shared state.
-1. The new reader can take over a segment by recording that it is doing so in the shared state.
-1. The new reader can start reading from the position it read from the shared state for the segment it picked up.
+## New Reader
+1. When a Reader joins a group its online status is added to the shared state.
```
Comment (RaulGracia) [inline documentation/src/docs/reader-group-design.md]: Why all the bullets are `1.`?

File: documentation/src/docs/reader-group-design.md [hunk 13]
```diff
@@ -9,85 +9,90 @@ You may obtain a copy of the License at
...
-    ReaderConfig cc = new ReaderConfig(props);
-    Reader<T> reader = a_stream.createReader("my_reader_id", "my_reader_group", cc);
+ClientFactory clientFactory = ClientFactory.withScope(scope, controllerURI);
+EventStreamReader< T > reader = clientFactory.createReader(readerId, READER_GROUP_NAME, serializer, readerConfig);
+
 ```
-When readers join the group they use the state to determine which segments to read from. When they shut down they update the state so that other readers can take over their segments.
+The Readers, while joining the group access the information stored on the state to determine which segments to read from. Once when they shut down, they update the state so that other Readers can take over their segments.
 
 # Failure detector
```
Comment (RaulGracia) [inline documentation/src/docs/reader-group-design.md]: I don't understand the structure of the documentation at this point. We start talking about `failure detector` but the associated text is not explaining it correctly. Even worse, a similar piece of content is repeated later on in `Reader going offline`. 

I suggest to refactor this part as follows:
1.- Replace `Failure detector` by something like `Reader Group Lifecycle: Example Operations`. Then, we can introduce the section with something like: `In what follows, we describe the workflow of the main operations executed during the liefecycle of a Reader Group.`
2.- Delete the content until `New Reader`.
3.- Then, each subsection subsection describes a different event within a Reader Group: `New Reader`, `Segments Merging`, `Reader Going Offline`. 

This structure may be extended to describe other Reader Group operations in the future.

File: documentation/src/docs/reader-group-design.md [hunk 14]
```diff
@@ -9,85 +9,90 @@ You may obtain a copy of the License at
...
 ## Segments merging
-1. When a reader comes to the end of its segment it records this information in the shared state.
-1. When all of the segments that are getting merged together are completed a reader may claim ownership of the following segment.
+1. When a Reader comes to the end of its segment it records this information in the shared state.
+1. When all of the segments that are getting merged together are completed, a Reader may claim ownership of the following segment.
+
+There is no ambiguity as to who the owner is, because it is stored in the shared state. There is no risk of a segment being ignored because every Reader can see the available segments by looking at the shared state and claim them.
+
+## Reader going offline
+1. When a Reader dies, the `readerOffline(readerId, position) api` method will be invoked either by the Reader itself in a graceful shutdown (internally to the close method) or via a "liveness detector". In either case the Reader's last position is written to the state.
```
Comment (RaulGracia) [inline documentation/src/docs/reader-group-design.md]: Maybe `api` should be API (without `this format`).

File: documentation/src/docs/reader-group-design.md [hunk 15]
```diff
@@ -9,85 +9,90 @@ You may obtain a copy of the License at
...
+1. Once the state has been updated by the new Reader, it is considered the owner of the segment and can read from it.
 
-There is no ambiguity as to who the owner is, because it is stored in the shared state. There is no risk of a segment being forgotten because any reader can see which segments are available by looking at the shared state and claim them.
 
-## Reader going offline.
-1. When a reader dies the readerOffline() method will be invoked either by the reader itself in a graceful shutdown (internally to the close method) or via a liveness detector. In either case the reader's last position is passed in.
-1. The last position is written to the state.
-1. Other readers will see this when they update their local state.
-1. Any of them can decide to take over one or more of the segments owned by the old reader from where it left off by recording their intention to do so in the state object.
-1. Once the state has been updated the new reader is considered the owner of the segment and can read from it at will.
```
Comment (RaulGracia) [inline documentation/src/docs/reader-group-design.md]: Here we can introduce another section called for example `Other Considerations on Reader Groups` to distinguish the two next subsections (which are not operations) from the previous ones.

File: documentation/src/docs/reader-group-design.md (hunk 7 above)
Comment (Aparnarr) [reply documentation/src/docs/reader-group-design.md]: @RaulGracia will decide on this post beta

File: documentation/src/docs/reader-group-design.md (hunk 8 above)
Comment (Aparnarr) [reply documentation/src/docs/reader-group-design.md]: accepted suggestion

File: documentation/src/docs/reader-group-design.md (hunk 9 above)
Comment (Aparnarr) [reply documentation/src/docs/reader-group-design.md]: Accepted suggestion

File: documentation/src/docs/reader-group-design.md [hunk 16]
```diff
@@ -11,83 +11,85 @@ You may obtain a copy of the License at
...
+Every time the Readers in a Reader Group change, the state can be updated. The replicated state is updated every time when one of the Readers starts reading from a new segment. Thus every Reader is aware of the segments owned by all the Readers in the their Reader Group.
 
 Given this information:
 
-- A new reader can infer which segments are available to read from. (By virtue of it being absent from the state)
-- Dealing with a segment being merged becomes easy, because the last reader to reach the end of its pre-merge segment knows it can freely take ownership of the new segment.
-- Readers can see their relative load and how they are progressing relative to the other readers in their group and can decide to transfer segments if things are out of balance.
-- This allows readers to take action directly to ensure all the events are read without the need for some external tracker.
+ - A new Reader can infer which segments are available to read from. (By virtue of it being absent from the state).
+ - Merging segments becomes easy, because the last Reader to reach the end of its pre-merge segment knows it can freely take ownership of the new segment.
```
[the remark this answers is not in the document]
Comment (Aparnarr) [reply documentation/src/docs/reader-group-design.md]: retained previous one!

File: documentation/src/docs/reader-group-design.md [hunk 17]
```diff
@@ -11,83 +11,85 @@ You may obtain a copy of the License at
...
+ - Readers can see their relative load and how they are progressing relative to the other Readers in their group and can decide to transfer segments if things are out of balance.
+ - This allows Readers to take action directly to ensure all the events are read without the need for some external tracker.
+
+## Reader Group APIs
+
+The external APIs to manage Reader Groups could be added to the `StreamManager` object. It consist of:
```
[the remark this answers is not in the document]
Comment (Aparnarr) [reply documentation/src/docs/reader-group-design.md]: Accepted suggestion

File: documentation/src/docs/reader-group-design.md (hunk 10 above)
Comment (Aparnarr) [reply documentation/src/docs/reader-group-design.md]: Accepted suggestion

File: documentation/src/docs/reader-group-design.md (hunk 13 above)
Comment (Aparnarr) [reply documentation/src/docs/reader-group-design.md]: @RaulGracia SO shoul d i remove the para starting with `We still need some effective mechanism`................till `exactly once processing guarantees.`

File: documentation/src/docs/reader-group-design.md (hunk 11 above)
Comment (Aparnarr) [reply documentation/src/docs/reader-group-design.md]: Accepted suggestion

File: documentation/src/docs/reader-group-design.md (hunk 12 above)
Comment (Aparnarr) [reply documentation/src/docs/reader-group-design.md]: not a issue, it renders the numbers sequentially in the output

File: documentation/src/docs/reader-group-design.md (hunk 14 above)
Comment (Aparnarr) [reply documentation/src/docs/reader-group-design.md]: Accepted suggestion

File: documentation/src/docs/reader-group-design.md (hunk 15 above)
Comment (Aparnarr) [reply documentation/src/docs/reader-group-design.md]: Accepted suggestion

File: documentation/src/docs/reader-group-design.md [hunk 18]
```diff
@@ -9,85 +9,90 @@ You may obtain a copy of the License at
...
 
-## ReaderGroup APIs
-The external APIs to manage ReaderGroups could be added to the StreamManager object. They consist of:
 ```
     ReaderGroup createReaderGroup(String name, Stream stream, ReaderGroupConfig config);
     ReaderGroup getReaderGroup(String name, Stream stream);
     void deleteReaderGroup(ReaderGroup group);
 ```
-When a ReaderGroup is created, it creates a [StateSynchronizer](state-synchronizer-design.md) shared by the readers. To join a ReaderGroup readers would just specify it in their configuration:
+When a Reader Group is created, it creates a [State Synchronizer](state-synchronizer-design.md) shared by the Readers. To join a Reader Group, Readers would just specify it in their configuration:
```
Comment (shrids) [inline documentation/src/docs/reader-group-design.md]: as I understand we should not have a space in between `State Synchronizer`. ref: https://github.com/pravega/pravega/blob/master/documentation/src/docs/state-synchronizer-design.md

nitpick : Also `Readers` should be changed to readers ( minus capital letter) right? 
We could follow the same pattern at other places too...

File: documentation/src/docs/reader-group-design.md [hunk 19]
```diff
@@ -9,85 +9,91 @@ You may obtain a copy of the License at
...
-    ReaderConfig cc = new ReaderConfig(props);
-    Reader<T> reader = a_stream.createReader("my_reader_id", "my_reader_group", cc);
+ClientFactory clientFactory = ClientFactory.withScope(scope, controllerURI);
+EventStreamReader< T > reader = clientFactory.createReader(readerId, READER_GROUP_NAME, serializer, readerConfig);
+
 ```
-When readers join the group they use the state to determine which segments to read from. When they shut down they update the state so that other readers can take over their segments.
+The Readers, while joining the group, access the information stored on the state to determine which segments to read from. Once when they shut down, they update the state so that other Readers can take over their segments.
 
 # Failure detector
```
Comment (shrids) [inline documentation/src/docs/reader-group-design.md]: My understanding of the Failure detector section is as follows.
- It should convey that the application should take care of tracking whether the `Reader`s are alive or not.
- It should convey that all the application should do is just invoke `ReaderGroup#readerOffline(String rReaderOffline(String ReaderId, Position lastPosition)`  on detecting a reader  failure ...
- The steps that need to be followed by the failure detector are simplified as it need not produce a view of the cluster or manage any state.

The `Examples` section talks about the operations that are performed during the three scenarios. (This implies that the section name should be changed from `Examples` to something meaningful that @RaulGracia  suggested.)
- New reader added
- Segments get merged
- Reader offline.

File: documentation/src/docs/reader-group-design.md (hunk 18 above)
Comment (Aparnarr) [reply documentation/src/docs/reader-group-design.md]: @shrids  The naming conventions, we are still working on ...as it is not consistent throughout all the 32 docs....
Hence i have planned to update the terminoloy.md  post beta. Is that sounds good!

File: documentation/src/docs/reader-group-design.md (hunk 19 above)
Comment (Aparnarr) [reply documentation/src/docs/reader-group-design.md]: ok

File: documentation/src/docs/reader-group-design.md [hunk 20]
```diff
@@ -9,85 +9,91 @@ You may obtain a copy of the License at
...
-1. Once the state has been updated the new reader is considered the owner of the segment and can read from it at will.
+# Other Considerations on Reader Groups
 
-## What happens if a reader does not keep up to date
-A Reader with out-of-date state can read from their existing segments without interference. The only disadvantage to this is that they will not shed load to another reader should one become available. However because they have to write to the shared state to start reading from any segment they don't already own, they must fetch up-to-date information before moving on to a new segment. 
+## What happens if a Reader does not keep up to date?
+A Reader with out-of-date state can read from their existing segments without interference. The only disadvantage to this is that they will not shed load to another Reader should one become available. However, because they have to write to the shared state to start reading from any segment which they don't already own, they must fetch up-to-date information before moving on to a new segment.
```
Comment (RaulGracia) [inline documentation/src/docs/reader-group-design.md]: `segment which they` -> I don't see the need to add `which` here.

File: documentation/src/docs/reader-group-design.md [hunk 21]
```diff
@@ -9,85 +9,91 @@ You may obtain a copy of the License at
...
+1. When all of the segments that are getting merged together are completed, a Reader may claim ownership of the following segment.
+
+There is no ambiguity as to who the owner is, because it is stored in the shared state. There is no risk of a segment being ignored because every Reader can see the available segments by looking at the shared state and claim them.
+
+## Reader going offline
+1. When a Reader dies, the `readerOffline(readerId, position)` API method will be invoked either by the Reader itself in a graceful shutdown (internally to the close method) or via a "liveness detector". In either case the Reader's last position is written to the state.
+
+1. If a null `Position` is sent then the last checkpointed position will be written to the state.
+1. This is used by the newer Readers when they take ownership of the segment(s) that were read by the older/offline reader.
+1. Any Reader can decide to take over one or more of the segments owned by the old Reader from where it left off by recording their intention to do so in the state object.
```
Comment (RaulGracia) [inline documentation/src/docs/reader-group-design.md]: various `segment` -> Stream Segment

File: documentation/src/docs/reader-group-design.md [hunk 22]
```diff
@@ -9,85 +9,91 @@ You may obtain a copy of the License at
...
+A consistent replicated state object representing the Reader Group metadata will be created in each Reader. This Reader Group metadata consists of:
 
-- a map of online readers to the segments they own
-- a list of positions in segments that can be taken over.
+ - A map of online Readers to the segments they own.
+ - A list of positions in segments that can be taken over.
 
-Every time the Readers in a ReaderGroup change, the state can be updated. Similarly each time one of the readers is going to start reading from a new segment, it can update the replicated state. 
-This allows all readers to know about all the other Readers in their ReaderGroup and which segments they own.
+Every time the Readers in a Reader Group change, the state can be updated. The replicated state is updated every time when one of the Readers starts reading from a new segment. Thus every Reader is aware of the segments owned by all the Readers in the their Reader Group.
```
Comment (RaulGracia) [inline documentation/src/docs/reader-group-design.md]: Please, change: `every time when one of the Readers starts reading from a new segment` -> `every time a Reader starts reading a new Segment.`

File: documentation/src/docs/reader-group-design.md (hunk 19 above)
Comment (Aparnarr) [reply documentation/src/docs/reader-group-design.md]: @raul @shrids  edited please check

File: documentation/src/docs/reader-group-design.md (hunk 21 above)
Comment (Aparnarr) [reply documentation/src/docs/reader-group-design.md]: ok will replace all segments withe Stream Segment..

File: documentation/src/docs/reader-group-design.md (hunk 22 above)
Comment (Aparnarr) [reply documentation/src/docs/reader-group-design.md]: done

**Comment by Aparnarr** [comment]
@raul addressed all ur comments:
Segment-> Stream Segment in all places
